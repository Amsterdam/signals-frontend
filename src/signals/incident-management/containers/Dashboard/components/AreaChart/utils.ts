// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import { Direction } from './ComparisonRate'
import type { ComparisonRateType } from './ComparisonRate'
import type { AreaChartValue as Value, Today } from '../../charts/types'

export const formatData = (values: Value[]): Value[] => {
  return values.map((value, i) =>
    i === values.length - 1
      ? {
          ...value,
          date: `${value.date} 12:00`,
          image: '/assets/images/area-chart-tooltip.svg',
        }
      : { ...value, date: `${value.date} 12:00` }
  )
}

export const getToday = (): Today => {
  const today = new Date()

  return {
    year: today.getFullYear(),
    // Vega lite does not work with zero indexed months
    month: today.getMonth() + 1,
    date: today.getDate(),
    hours: 12,
    minutes: 0o0,
  }
}

export const getPercentage = (values: Value[]): ComparisonRateType => {
  const thisWeek = values.reduce((total, value) => total + value.amount, 0)
  const lastWeek = values.reduce(
    (total, value) => total + value.amount_week_earlier,
    0
  )

  return {
    direction: thisWeek < lastWeek ? Direction.DOWN : Direction.UP,
    percentage: Math.abs(Math.round((thisWeek / lastWeek - 1) * 100)),
  }
}
