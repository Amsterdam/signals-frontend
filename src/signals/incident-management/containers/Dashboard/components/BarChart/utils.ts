// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import configuration from 'shared/services/configuration/configuration'

import type { RawData } from './types'
import { constants } from '../../charts'
import type { BarChartValue } from '../../charts'

export const getQueryList = (queryString: string) => {
  const endpoint = configuration.INCIDENT_PRIVATE_ENDPOINT
  return constants.statusList.map((status) => {
    const params = queryString
      ? `${queryString}&status=${status.query}`
      : `status=${status.query}`
    return `${endpoint}stats/total?${params}`
  })
}

export const formatData = (rawData: RawData[]): BarChartValue[] => {
  return rawData.map((value, i) => {
    return {
      status: constants.statusList[i].label,
      nrOfIncidents: value.total,
    }
  })
}

export const getTotalNrOfIncidents = (rawData: RawData[]): number => {
  return rawData.reduce(
    (aggregatedTotal, value) => aggregatedTotal + value.total,
    0
  )
}
