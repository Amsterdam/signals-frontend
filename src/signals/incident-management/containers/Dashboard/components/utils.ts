// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { IncidentCount } from '../charts/types'

export const getMaxDomain = (values: IncidentCount[]) =>
  Math.max(...values.map((v) => v.amount)) * 1.2
