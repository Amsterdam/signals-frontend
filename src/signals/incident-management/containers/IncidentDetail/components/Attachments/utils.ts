// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam
import type { Attachment } from '../../types'

export const getAttachmentFileName = (location: Attachment['location']) =>
  location.split('?')[0].split('/').pop()
