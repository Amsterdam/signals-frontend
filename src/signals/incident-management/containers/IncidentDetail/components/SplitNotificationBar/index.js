import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import { Button } from '@datapunt/asc-ui';
import styled, { ascDefaultTheme } from '@datapunt/asc-core';
import {
  Close,
} from '@datapunt/asc-assets';

const StyledWrapper = styled.div`
  flex-basis: 100%;
  color: ${ascDefaultTheme.colors.bright.main};
  a {
    color: ${ascDefaultTheme.colors.bright.main};
  }
  `;

const StyledSuccess = styled.div`
  background-color: #00a03c;
  padding: 6px 10px;
`;

const StyledError = styled.div`
  background-color: ${ascDefaultTheme.colors.error.main};
  padding: 6px 10px;
`;

const StyledCloseButton = styled(Button)`
  float: right;
  background-color: transparent;

  &:hover {
    background-color: transparent;
  }

  svg path {
    fill: ${ascDefaultTheme.colors.bright.main};
    ;
  }
`;

const getErrorMessage = (status) => {
  switch (status) {
    case 403:
      return 'U bent niet bevoegd om deze melding te splitsen.';

    case 412:
      return 'U kunt geen meldingen splitsen die al gesplitst zijn.';

    default:
      return 'Er is een onbekende fout ontstaan.';
  }
};

const getDelimiter = (key, length) => {
  if (key === length - 2) {
    return ' en ';
  }
  return key < length - 1 ? ', ' : '';
};

const renderCloseButton = (onDismissSplitNotification) => (
  <StyledCloseButton
    size={20}
    variant="blank"
    icon={<Close />}
    data-testid="splitNotificationBarCloseButton"
    onClick={onDismissSplitNotification}
  />
);

const SplitNotificationBar = ({ data, onDismissSplitNotification }) => (
  <StyledWrapper data-testid="splitNotificationBar">
    {data && data.id && data.created && data.created.children && Array.isArray(data.created.children) ?
      <StyledSuccess>
        {renderCloseButton(onDismissSplitNotification)}

        Melding {data.id} is gesplitst in&nbsp;
        {data.created.children.map((item, key) =>
        (<span key={item.id}>
          <NavLink to={`/manage/incident/${item.id}`}>{item.id}</NavLink>
          {getDelimiter(key, data.created.children.length)}
        </span>)
      )}
      </StyledSuccess>
      : ''}
    {data && data.response && data.response.status ?
      <StyledError>
        {renderCloseButton(onDismissSplitNotification)}

        De melding is helaas niet gesplitst.&nbsp;
        {getErrorMessage(data.response.status)}
      </StyledError>
        : ''}
  </StyledWrapper>
);

SplitNotificationBar.propTypes = {
  data: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.object,
    PropTypes.array
  ]),
  onDismissSplitNotification: PropTypes.func.isRequired
};

export default SplitNotificationBar;
