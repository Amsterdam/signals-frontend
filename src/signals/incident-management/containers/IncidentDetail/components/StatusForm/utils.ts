// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2020 - 2021 Gemeente Amsterdam
import type { AlertLevel } from '@amsterdam/asc-ui'

import configuration from 'shared/services/configuration/configuration'
import {
  changeStatusOptionList,
  isStatusClosed,
} from 'signals/incident-management/definitions/statusList'
import { StatusCode } from 'types/status-code'

import * as constants from './constants'

export const emailSentWhenStatusChangedTo = ({
  toStatus,
  fromStatus,
  isSplitIncident,
}: {
  toStatus: StatusCode
  fromStatus: StatusCode
  isSplitIncident: boolean
}): boolean => {
  if (isSplitIncident) return false

  if (
    !configuration.featureFlags.reporterMailHandledNegativeContactEnabled &&
    fromStatus === StatusCode.VerzoekTotHeropenen &&
    toStatus === StatusCode.Afgehandeld
  ) {
    return false
  }

  return Boolean(
    changeStatusOptionList.find(
      ({ email_sent_when_set, key }) => email_sent_when_set && toStatus === key
    )
  )
}

export const textIsRequired = ({
  fromStatus,
  toStatus,
  isSplitIncident,
}: {
  fromStatus: StatusCode
  toStatus: StatusCode
  isSplitIncident: boolean
}): boolean => {
  if (
    isSplitIncident ||
    !configuration.featureFlags.reporterMailHandledNegativeContactEnabled
  ) {
    return [
      StatusCode.Afgehandeld,
      StatusCode.Ingepland,
      StatusCode.Heropend,
      StatusCode.ReactieGevraagd,
    ].includes(toStatus)
  } else {
    return emailSentWhenStatusChangedTo({
      fromStatus,
      toStatus,
      isSplitIncident,
    })
  }
}

type Warning = {
  key: string
  level: AlertLevel
  content: string
  heading?: string
}

export const determineWarnings = ({
  originalStatus,
  toStatus,
  isSplitIncident,
  hasEmail,
  hasOpenChildren,
  categorySlug,
}: {
  originalStatus: StatusCode
  toStatus: StatusCode
  isSplitIncident: boolean
  hasEmail: boolean
  hasOpenChildren: boolean
  categorySlug?: string
}): Warning[] => {
  const warnings: Warning[] = []

  if (isSplitIncident) return []

  if (isStatusClosed(toStatus) && hasOpenChildren) {
    warnings.push({
      key: 'has-open-child-incidents-warning',
      heading: constants.DEELMELDINGEN_STILL_OPEN_HEADING,
      content: constants.DEELMELDINGEN_STILL_OPEN_CONTENT,
      level: 'info',
    })
  }

  if (originalStatus === StatusCode.ReactieGevraagd && hasEmail) {
    warnings.push({
      key: 'has-open-reply-request-warning',
      heading: constants.REPLY_CHANGE_STATUS_HEADING,
      content: constants.REPLY_CHANGE_STATUS_CONTENT,
      level: 'info',
    })
  }

  if (
    configuration.featureFlags.disableClosingCategoryOverigOverig &&
    toStatus === StatusCode.Afgehandeld &&
    categorySlug === 'overig'
  ) {
    warnings.push({
      key: 'has-category-overig',
      heading: constants.CATEGORY_OVERIG_HEADING,
      content: constants.CATEGORY_OVERIG_CONTENT,
      level: 'error',
    })
  } else if (toStatus === StatusCode.Afgehandeld)
    warnings.push({
      key: 'end-status-warning',
      content: constants.AFGEHANDELD_CONTENT,
      level: 'neutral',
    })

  if (toStatus === StatusCode.ReactieGevraagd && !hasEmail) {
    warnings.push({
      key: 'has-no-email-reply-warning',
      heading: constants.REPLY_NO_MAIL_HEADING,
      content: constants.REPLY_NO_MAIL_CONTENT,
      level: 'error',
    })
  }

  return warnings
}

export const getTextConfig = (statusCode: StatusCode) => {
  return statusCode === StatusCode.ReactieGevraagd
    ? {
        label: constants.REPLY_MAIL_LABEL,
        maxLength: constants.REPLY_MAIL_MAX_LENGTH,
        rows: 6,
      }
    : {
        label: constants.DEFAULT_TEXT_LABEL,
        maxLength: constants.DEFAULT_TEXT_MAX_LENGTH,
        rows: 9,
      }
}
