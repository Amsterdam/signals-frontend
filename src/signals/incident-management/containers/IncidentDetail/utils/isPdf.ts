// SPDX-License-Identifier: MPL-2.0
// Copyright (C) 2023 Gemeente Amsterdam, Vereniging van Nederlandse Gemeenten
export const isPdf = (url: string) => {
  return !!url.match(/[a-zA-Z0-9-_/]+\.pdf/)
}
