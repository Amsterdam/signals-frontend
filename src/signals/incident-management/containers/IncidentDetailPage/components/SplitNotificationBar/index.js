import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';

import './style.scss';

const SplitNotificationBar = ({ payload, onClose }) => (
  <div>
    {payload && payload.id && payload.created && Array.isArray(payload.created) ?
      <div className="split-notification-bar">
        <div className="split-notification-bar__body">
          <button className="split-notification-bar__close-button" onClick={onClose} />

          Melding {payload.id} is gesplitst in
          {payload.created.map((item) =>
            (<span key={item.id} className="split-notification-bar__link">
              &nbsp;<NavLink to={`/manage/incident/${item.id}`}>{item.id}</NavLink>
            </span>)
          )}
        </div>
      </div>
      : ''}
  </div>
);

SplitNotificationBar.propTypes = {
  payload: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.object,
    PropTypes.array
  ]),
  onClose: PropTypes.func.isRequired
};

export default SplitNotificationBar;
