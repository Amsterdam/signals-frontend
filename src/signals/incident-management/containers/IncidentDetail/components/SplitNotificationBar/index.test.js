import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { withAppContext } from 'test/utils';

import SplitNotificationBar from './index';

describe('<SplitNotificationBar />', () => {
  let props;

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('success', () => {
    beforeEach(() => {
      props = {
        data: {
          id: '42',
          created: {
            children: [
              { id: 43 },
              { id: 44 }
            ]
          }
        },
        onDismissSplitNotification: jest.fn()
      };
    });

    it('should render 2  items correctly', () => {
      const { queryByTestId } = render(
        withAppContext(<SplitNotificationBar {...props} />)
      );

      expect(queryByTestId('splitNotificationBar')).toHaveTextContent(/^Melding 42 is gesplitst in 43 en 44$/);
    });

    it('should render 3 items correctly', () => {
      props.data.created.children.push({ id: 45 });
      const { queryByTestId } = render(
        withAppContext(<SplitNotificationBar {...props} />)
      );

      expect(queryByTestId('splitNotificationBar')).toHaveTextContent(/^Melding 42 is gesplitst in 43, 44 en 45$/);
    });

    it('should render no items correctly', () => {
      props.data.created = undefined;
      const { queryByTestId } = render(
        withAppContext(<SplitNotificationBar {...props} />)
      );

      expect(queryByTestId('splitNotificationBar')).toHaveTextContent(/^$/);
    });

    it('should dissmiss bar when clicked', () => {
      const { queryByTestId } = render(
        withAppContext(<SplitNotificationBar {...props} />)
      );
      fireEvent.click(queryByTestId('splitNotificationBarCloseButton'));

      expect(props.onDismissSplitNotification).toHaveBeenCalledTimes(1);
    });
  });

  describe('error', () => {
    beforeEach(() => {
      props = {
        data: {
          response: {
            status: 503
          }
        },
        onDismissSplitNotification: jest.fn()
      };
    });

    it('should render general error correctly', () => {
      const { queryByTestId } = render(
        withAppContext(<SplitNotificationBar {...props} />)
      );

      expect(queryByTestId('splitNotificationBar')).toHaveTextContent(/^De melding is helaas niet gesplitst\. Er is een onbekende fout ontstaan\.$/);
    });

    it('should render general 403 correctly', () => {
      props.data.response.status = 403;
      const { queryByTestId } = render(
        withAppContext(<SplitNotificationBar {...props} />)
      );

      expect(queryByTestId('splitNotificationBar')).toHaveTextContent(/^De melding is helaas niet gesplitst\. U bent niet bevoegd om deze melding te splitsen\.$/);
    });

    it('should render general 412 correctly', () => {
      props.data.response.status = 412;
      const { queryByTestId } = render(
        withAppContext(<SplitNotificationBar {...props} />)
      );

      expect(queryByTestId('splitNotificationBar')).toHaveTextContent(/^De melding is helaas niet gesplitst\. U kunt geen meldingen splitsen die al gesplitst zijn\.$/);
    });

    it('should dissmiss bar when clicked', () => {
      const { queryByTestId } = render(
        withAppContext(<SplitNotificationBar {...props} />)
      );
      fireEvent.click(queryByTestId('splitNotificationBarCloseButton'));

      expect(props.onDismissSplitNotification).toHaveBeenCalledTimes(1);
    });
  });
});
