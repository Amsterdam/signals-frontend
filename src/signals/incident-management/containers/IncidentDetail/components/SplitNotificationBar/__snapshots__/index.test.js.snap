// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SplitNotificationBar /> error should render general 403 correctly 1`] = `
<div>
  <div
    className="split-notification-bar error"
  >
    <div
      className="split-notification-bar__body"
    >
      <button
        className="split-notification-bar__close-button"
        onClick={[MockFunction]}
      />
      De melding is helaas niet gesplitst. 
      U bent niet bevoegd om deze melding te splitsen.
    </div>
  </div>
</div>
`;

exports[`<SplitNotificationBar /> error should render general 412 correctly 1`] = `
<div>
  <div
    className="split-notification-bar error"
  >
    <div
      className="split-notification-bar__body"
    >
      <button
        className="split-notification-bar__close-button"
        onClick={[MockFunction]}
      />
      De melding is helaas niet gesplitst. 
      U kunt geen meldingen splitsen die al gesplitst zijn.
    </div>
  </div>
</div>
`;

exports[`<SplitNotificationBar /> error should render general error correctly 1`] = `
<div>
  <div
    className="split-notification-bar error"
  >
    <div
      className="split-notification-bar__body"
    >
      <button
        className="split-notification-bar__close-button"
        onClick={[MockFunction]}
      />
      De melding is helaas niet gesplitst. 
      Er is een onbekende fout ontstaan.
    </div>
  </div>
</div>
`;

exports[`<SplitNotificationBar /> success should render 2  items correctly 1`] = `
<div>
  <div
    className="split-notification-bar success"
  >
    <div
      className="split-notification-bar__body"
    >
      <button
        className="split-notification-bar__close-button"
        onClick={[MockFunction]}
      />
      Melding 
      42
       is gesplitst in
      <span
        className="split-notification-bar__link"
        key="43"
      >
         
        <NavLink
          to="/manage/incident/43"
        >
          43
        </NavLink>
      </span>
      <span
        className="split-notification-bar__link"
        key="44"
      >
         
        <NavLink
          to="/manage/incident/44"
        >
          44
        </NavLink>
      </span>
    </div>
  </div>
</div>
`;

exports[`<SplitNotificationBar /> success should render 3 items correctly 1`] = `
<div>
  <div
    className="split-notification-bar success"
  >
    <div
      className="split-notification-bar__body"
    >
      <button
        className="split-notification-bar__close-button"
        onClick={[MockFunction]}
      />
      Melding 
      42
       is gesplitst in
      <span
        className="split-notification-bar__link"
        key="43"
      >
         
        <NavLink
          to="/manage/incident/43"
        >
          43
        </NavLink>
      </span>
      <span
        className="split-notification-bar__link"
        key="44"
      >
         
        <NavLink
          to="/manage/incident/44"
        >
          44
        </NavLink>
      </span>
      <span
        className="split-notification-bar__link"
        key="45"
      >
         
        <NavLink
          to="/manage/incident/45"
        >
          45
        </NavLink>
      </span>
    </div>
  </div>
</div>
`;

exports[`<SplitNotificationBar /> success should render no items correctly 1`] = `<div />`;
